### PR TITLE
feat: shopify sessionid stitching

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

> This pull request introduces an enhancement of **sessionId stitching** between Shopify Client-Side and Server-Side events functionality by leveraging the rudderSessionIdentifier event and using of Redis for storing the `cartToken` versus `SessionId` mapping.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/sources/shopify/transform.js [hunk 1]
```diff
@@ -171,8 +173,10 @@ const processEvent = async (inputEvent, metricMetadata) => {
   }
   if (message.type !== EventType.IDENTIFY) {
     let anonymousId;
+    let sessionId;
     if (useRedisDatabase) {
       anonymousId = await getAnonymousIdFromDb(message, metricMetadata);
```
Comment (koladilip) [inline src/v0/sources/shopify/transform.js:178]: not possible to get both anonid and session id in one call?

**Review by koladilip — APPROVED** [review]

File: src/v0/sources/shopify/transform.js [hunk 2]
```diff
@@ -198,15 +205,31 @@ const processEvent = async (inputEvent, metricMetadata) => {
...
+      /*
+      cart_token: {
+      sessionId: "session_id1"
+     }
+     */
+    }
     try {
       await RedisDB.setVal(`${event.cartToken}`, value);
     } catch (e) {
+      logger.debug(`{{SHOPIFY::}} cartToken maping set call Failed due redis error ${e}`);
```
Comment (mihir-4116) [inline src/v0/sources/shopify/transform.js:232]: `logger.debug(`{{SHOPIFY::}} Failed to set cartToken mapping due to a Redis error: ${e}`);`

File: src/v0/sources/shopify/util.js [hunk 3]
```diff
@@ -126,6 +126,7 @@ const getAnonymousIdFromDb = async (message, metricMetadata) => {
     try {
       anonymousId = await RedisDB.getVal(`${cartToken}`, 'anonymousId');
     } catch (e) {
+      logger.debug(`{{SHOPIFY::}} anonymousId get call Failed due redis error ${e}`);
```
Comment (mihir-4116) [inline src/v0/sources/shopify/util.js:129]: `logger.debug(`{{SHOPIFY::}} Failed to retrieve anonymousId from Redis due to an error: ${e}`);`

File: src/v0/sources/shopify/util.js [hunk 4]
```diff
@@ -146,6 +147,41 @@ const getAnonymousIdFromDb = async (message, metricMetadata) => {
...
+  const cartToken = getCartToken(message);
+  const { event } = message;
+  if (!isDefinedAndNotNull(cartToken)) {
+    return null;
+  }
+  let sessionId;
+  try {
+    sessionId = await RedisDB.getVal(`${cartToken}`, 'sessionId');
+  } catch (e) {
+    logger.debug(`Time: ${new Date()} {{SHOPIFY::}} sessionId get call Failed due redis error ${e}`);
```
Comment (mihir-4116) [inline src/v0/sources/shopify/util.js:165]: logger.debug(`Time: ${new Date()} {{SHOPIFY::}} Failed to retrieve sessionId from Redis due to an error: ${e}`);

File: src/v0/sources/shopify/util.js [hunk 5]
```diff
@@ -158,6 +194,7 @@ const updateCartItemsInRedis = async (cartToken, newCartItemsHash, metricMetadat
   try {
     await RedisDB.setVal(`${cartToken}`, value);
   } catch (e) {
+    logger.debug(`{{SHOPIFY::}} itemsHash set call Failed due redis error ${e}`);
```
Comment (mihir-4116) [inline src/v0/sources/shopify/util.js:197]: `logger.debug(`{{SHOPIFY::}} Failed to set itemsHash due to a Redis error: ${e}`);`

File: src/v0/sources/shopify/util.js [hunk 6]
```diff
@@ -176,6 +213,7 @@ const checkAndUpdateCartItems = async (inputEvent, metricMetadata) => {
       });
     }
   } catch (e) {
+    logger.debug(`{{SHOPIFY::}} itemsHash get call Failed due redis error ${e}`);
```
Comment (mihir-4116) [inline src/v0/sources/shopify/util.js:216]: `logger.debug(`{{SHOPIFY::}} Failed to retrieve itemsHash from Redis due to an error: ${e}`);`

File: src/v0/sources/shopify/util.js [hunk 7]
```diff
@@ -146,6 +147,41 @@ const getAnonymousIdFromDb = async (message, metricMetadata) => {
...
+    stats.increment('shopify_redis_failures', {
+      type: 'get',
+      ...metricMetadata,
+    });
+  }
+  stats.increment('shopify_redis_calls', {
+    type: 'get',
+    ...metricMetadata,
+  });
+  if (sessionId === null) {
```
Comment (mihir-4116) [inline src/v0/sources/shopify/util.js:176]: is you create a common isNull function or check if it's available then use it here?

**Review by mihir-4116 — CHANGES_REQUESTED** [review]

**Comment by anantjain45823** [comment]
Moved to #2204

**Pull request closed without merging**